File: app/src/main/java/com/armsnyder/leftorright/MainActivity.java
package com.armsnyder.leftorright;

import android.content.res.ColorStateList;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Locale;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";
    private SwipeRefreshLayout swipeRefreshLayout;
    private ImageView arrowImageView;
    private LeftOrRightIO leftOrRightIO;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        swipeRefreshLayout = findViewById(R.id.swipe2);
        swipeRefreshLayout.setOnRefreshListener(this::refresh);
        arrowImageView = findViewById(R.id.imageView);
        Toolbar myToolbar = findViewById(R.id.toolbar2);
        setSupportActionBar(myToolbar);
        leftOrRightIO = new LeftOrRightIO(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        refresh();
    }

    @Override
    protected void onPause() {
        super.onPause();
        hideArrow();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_refresh:
                refresh();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    public void refresh() {
        if (!swipeRefreshLayout.isRefreshing()) {
            swipeRefreshLayout.setRefreshing(true);
        }
        hideArrow();
        leftOrRightIO.getAdvice().whenComplete((advice, err) -> {
            if (err == null) {
                Direction direction = advice.getDirection();
                arrowImageView.setImageResource(directionToImageResource(direction));
                arrowImageView.setContentDescription(direction.toString().toLowerCase());
                TextView etaView = findViewById(R.id.eta);
                etaView.setText(String.format(Locale.ENGLISH, "%d min", advice.getMinutesUntilArrival()));
                TextView routeView = findViewById(R.id.route);
                routeView.setText(advice.getRoute());
                showArrow();
            } else {
                Log.e(TAG, err.getMessage(), err);
                Snackbar.make(findViewById(R.id.container2), R.string.generic_error,
                        Snackbar.LENGTH_LONG).show();
            }
            swipeRefreshLayout.setRefreshing(false);
        });
    }

    private void showArrow() {
        TextView etaView = findViewById(R.id.eta);
        etaView.setTextColor(getResources().getColor(R.color.active, null));
        ImageView imageView = findViewById(R.id.imageView);
        imageView.setImageTintList(ColorStateList.valueOf(getResources().getColor(R.color.active, null)));
    }

    private void hideArrow() {
        TextView etaView = findViewById(R.id.eta);
        etaView.setTextColor(getResources().getColor(R.color.plain, null));
        ImageView imageView = findViewById(R.id.imageView);
        imageView.setImageTintList(ColorStateList.valueOf(getResources().getColor(R.color.plain, null)));
    }

    private int directionToImageResource(Direction direction) {
        switch (direction) {
            case LEFT:
                return R.drawable.ic_arrow_back_24dp;
            case RIGHT:
                return R.drawable.ic_arrow_forward_24dp;
            default:
                throw new IllegalStateException();
        }
    }
}

File: app/src/main/java/com/armsnyder/leftorright/MainActivity.kt
package com.armsnyder.leftorright

import android.content.res.ColorStateList
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.launch
import java.util.*

class MainActivity : AppCompatActivity() {
    private val leftOrRightIO: LeftOrRightIO by lazy { LeftOrRightIO(this) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        swipeLayout.setOnRefreshListener(this::refresh)
        setSupportActionBar(toolbar)
    }

    override fun onResume() {
        super.onResume()
        refresh()
    }

    override fun onPause() {
        super.onPause()
        hideArrow()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_refresh -> {
                refresh()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun refresh() {
        if (!swipeLayout.isRefreshing) {
            swipeLayout.isRefreshing = true
        }
        hideArrow()
        launch(UI) {
            try {
                val (direction, minutesUntilArrival, route) = leftOrRightIO.getAdvice()
                arrowImage.setImageResource(directionToImageResource(direction))
                arrowImage.contentDescription = direction.toString().toLowerCase()
                etaText.text = String.format(Locale.ENGLISH, "%d min", minutesUntilArrival)
                routeText.text = route
                showArrow()
            } catch (e: Exception) {
                Log.e(TAG, e.message, e)
                Snackbar.make(coordinatorLayout, R.string.generic_error, Snackbar.LENGTH_LONG)
                        .show()
            }
            swipeLayout.isRefreshing = false
        }
    }

    private fun showArrow() {
        etaText.setTextColor(resources.getColor(R.color.active, null))
        arrowImage.imageTintList = ColorStateList.valueOf(resources.getColor(R.color.active, null))
    }

    private fun hideArrow() {
        etaText.setTextColor(resources.getColor(R.color.plain, null))
        arrowImage.imageTintList = ColorStateList.valueOf(resources.getColor(R.color.plain, null))
    }

    private fun directionToImageResource(direction: Direction): Int {
        return when (direction) {
            Direction.LEFT -> R.drawable.ic_arrow_back_24dp
            Direction.RIGHT -> R.drawable.ic_arrow_forward_24dp
        }
    }

    companion object {
        private const val TAG = "MainActivity"
    }
}
